src/stream/src/executor/source/kafka_backfill_executor.rs

```diff
@@ -0,0 +1,895 @@
+                    }
+                    yield Message::Barrier(barrier);
+                }
+                Message::Chunk(chunk) => {
+                    yield Message::Chunk(chunk);
+                }
+                Message::Watermark(watermark) => {
+                    yield Message::Watermark(watermark);
+                }
+            }
+        }
+    }
+
+    /// All splits finished backfilling.
+    ///
+    /// We check all splits for the source, including other actors' splits here, before going to the forward stage.
+    /// Otherwise if we break early, but after rescheduling, an unfinished split is migrated to
+    /// this actor, we still need to backfill it.
+    async fn backfill_finished(&self, states: &BackfillStates) -> StreamExecutorResult<bool> {
+        Ok(states
+            .values()
+            .all(|state| matches!(state, BackfillState::Finished))
+            && self
+                .backfill_state_store
+                .scan()
+                .await?
+                .into_iter()
+                .all(|state| matches!(state, BackfillState::Finished)))
+    }
+
+    /// For newly added splits, we do not need to backfill and can directly forward from upstream.
+    async fn apply_split_change(
+        &mut self,
+        split_assignment: &HashMap<ActorId, Vec<SplitImpl>>,
+        stage: &mut BackfillStage,
+        should_trim_state: bool,
+    ) -> StreamExecutorResult<bool> {
+        self.metrics
+            .source_split_change_count
+            .with_label_values(
+                &self
+                    .get_metric_labels()
+                    .iter()
+                    .map(AsRef::as_ref)
+                    .collect::<Vec<&str>>(),
+            )
+            .inc();
+        if let Some(target_splits) = split_assignment.get(&self.actor_ctx.id).cloned() {
+            if self
+                .update_state_if_changed(target_splits, stage, should_trim_state)
+                .await?
+            {
+                // Note: we don't rebuild backfill_stream here, due to some complex lifetime issues.
+                return Ok(true);
+            }
+        }
+
+        Ok(false)
+    }
+
+    /// Returns `true` if split changed. Otherwise `false`.
+    async fn update_state_if_changed(
+        &mut self,
+        target_splits: Vec<SplitImpl>,
+        stage: &mut BackfillStage,
+        should_trim_state: bool,
+    ) -> StreamExecutorResult<bool> {
+        let target_splits: HashMap<_, _> = target_splits
+            .into_iter()
+            .map(|split| (split.id(), split))
+            .collect();
+
+        let mut target_state: HashMap<SplitId, BackfillState> =
+            HashMap::with_capacity(target_splits.len());
+
+        let mut split_changed = false;
+
+        // Checks added splits
+        for (split_id, mut split) in target_splits {
+            if let Some(s) = stage.states.get(&split_id) {
+                target_state.insert(split_id, s.clone());
```

Comment:
> For existing splits, we should use the latest offset from the cache. `target_splits` is from meta and contains the initial offset.

I'm unsure whether this is always correct when scaling. The ideas are similar to the documentation on the function [`cache_may_stale`](https://github.com/risingwavelabs/risingwave/blob/25219d51be5232c22e34b607535f0ae6440fdc67/src/stream/src/cache/mod.rs#L44). The safest way should be always recovering states from the state store.

Reply:
Do you mean the logic in SourceExecutor is also problematic?

Reply:
I guess so. 😄

Reply:
https://github.com/risingwavelabs/risingwave/issues/15591 Then let's fix it later